tools/test-case-generators/genall.go

```diff
@@ -0,0 +1,469 @@
+	solver := dnfjson.NewSolver(d.ModulePlatformID(), d.Releasever(), arch, cacheDir)
+	solver.SetDNFJSONPath("./dnf-json")
+	packageSets := imageType.PackageSets(bp, repos)
+	depsolvedSets := make(map[string][]rpmmd.PackageSpec)
+	for name, pkgSet := range packageSets {
+		res, err := solver.Depsolve(pkgSet)
+		if err != nil {
+			return nil, err
+		}
+		depsolvedSets[name] = res.Dependencies
+	}
+	return depsolvedSets, nil
+}
+
+func save(manifest distro.Manifest, pkgs map[string][]rpmmd.PackageSpec, cr composeRequest, path, filename string) {
+	data := struct {
+		ComposeRequest composeRequest                 `json:"compose-request"`
+		Manifest       distro.Manifest                `json:"manifest"`
+		RPMMD          map[string][]rpmmd.PackageSpec `json:"rpmmd"`
+		NoImageInfo    bool                           `json:"no-image-info"`
+	}{
+		cr, manifest, pkgs, true,
+	}
+	b, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to marshal data for %q: %s\n", filename, err.Error())
+	}
+	b = append(b, '\n') // add new line at end of file
+	fpath := filepath.Join(path, filename)
+	fp, err := os.Create(fpath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create output file %q: %s\n", fpath, err.Error())
+		return
+	}
+	defer fp.Close()
+	if _, err := fp.Write(b); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to write output file %q: %s\n", fpath, err.Error())
+		return
+	}
+}
+
+func filterRepos(repos []repository, typeName string) []repository {
+	filtered := make([]repository, 0)
+	for _, repo := range repos {
+		if len(repo.ImageTypeTags) == 0 {
+			filtered = append(filtered, repo)
+		} else {
+			for _, tt := range repo.ImageTypeTags {
+				if tt == typeName {
+					filtered = append(filtered, repo)
+					break
+				}
+			}
+		}
+	}
+	return filtered
+}
+
+// collects requests from a formatRequestMap based on image type
+func requestsByImageType(requestMap formatRequestMap) map[string]map[string]manifestRequest {
+	imgTypeRequestMap := make(map[string]map[string]manifestRequest)
+
+	for name, req := range requestMap {
+		it := req.ComposeRequest.ImageType
+		reqs := imgTypeRequestMap[it]
+		if reqs == nil {
+			reqs = make(map[string]manifestRequest)
+		}
+		reqs[name] = req
+		imgTypeRequestMap[it] = reqs
+	}
+	return imgTypeRequestMap
+}
+
+type workerQueue struct {
+	jobQueue chan manifestJob
+	msgQueue chan string
+	errQueue chan error
+	errors   []error
+	nworkers uint
+	njobs    uint
+	ndone    uint
+}
+
+func newWorkerQueue(nworkers uint, njobs uint) *workerQueue {
+	wq := workerQueue{
+		jobQueue: make(chan manifestJob, njobs),
+		msgQueue: make(chan string, nworkers),
+		errQueue: make(chan error, nworkers),
+		errors:   make([]error, 0, nworkers),
+		nworkers: nworkers,
+		njobs:    njobs,
+		ndone:    0,
+	}
+
+	return &wq
+}
+
+func (wq *workerQueue) start() {
+	for idx := uint(0); idx < wq.nworkers; idx++ {
+		go wq.startWorker()
```

Comment:
This has the potential of leaving dangling goroutines behind, what about adding a `sync.WaitGroup` and wait for it in `finish`?